### PR TITLE
Add a note about arrow crate security / safety

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

Closes https://github.com/apache/arrow-rs/issues/627

# Rationale for this change
 
If not used carefully, unsafe code can be written using arrow-rs, so we should tell users about this so they can be well informed
See previous discussion on the mailing list:
* https://lists.apache.org/thread.html/recac1f6dc982bab2923f8fb6992e2d4c927f46daff5f03ed6c4de19c%40%3Cdev.arrow.apache.org%3E
* https://lists.apache.org/thread.html/r8b76b9f6d5cf8624627dcefe74b093a9f32746645e81fe0c9fc45edc%40%3Cdev.arrow.apache.org%3E
*  https://lists.apache.org/thread.html/rfbef397ab13bf6b120a9e9d40446809df5fa8e7b783681223f2cc6e0%40%3Cdev.arrow.apache.org%3E

# What changes are included in this PR?
Update README in arrow crate explaining what is going on

### CODE REVIEW

File: arrow/README.md [hunk 1]
```diff
@@ -35,6 +39,26 @@ The arrow crate provides the following optional features:
   implementations of some [compute](https://github.com/apache/arrow/tree/master/rust/arrow/src/compute)
   kernels using explicit SIMD processor intrinsics.
 
+## Safety
+
+TLDR: You should avoid using the `alloc` and `buffer` and `bitmap` modules if at all possible. These modules contain `unsafe` code and are easy to misuse.
```
[the remark this answers is not in the document]
Comment (nevi-me) [reply arrow/README.md:44]: Yeah, I think it is, and I like your suggestion of putting them behind a feature flag

File: arrow/README.md [hunk 2]
```diff
@@ -35,6 +39,26 @@ The arrow crate provides the following optional features:
...
+
+```rust
+let buffer = Buffer::from_slic_ref(&[0i32, 2i32])
+let data = ArrayData::new(DataType::Int64, 10, 0, None, 0, vec![buffer], vec![]);
+let array = Float64Array::from(Arc::new(data));
+
+println!("{:?}", array.value(1));
+```
+
+NOTE: We plan to deprecate and make these modules private as part of a follow on release, as part of our journey of redesigning this crate.
```
[the remark this answers is not in the document]
Comment (nevi-me) [reply arrow/README.md:60]: Given that the change could be landed in a short period, I would omit the note, as it would be temporary.

**Review by nevi-me — APPROVED** [review]
One comment about possibly removing a note

**Review by Dandandan — APPROVED** [review]

File: arrow/README.md [hunk 3]
```diff
@@ -35,6 +39,24 @@ The arrow crate provides the following optional features:
...
 
+## Safety
+
+TLDR: You should avoid using the `alloc` and `buffer` and `bitmap` modules if at all possible. These modules contain `unsafe` code and are easy to misuse.
+
+As with all open source code, you should carefully evaluate the suitability of `arrow` for your project, taking into consideration your needs and risk tolerance prior to use.
+
+_Background_: There are various parts of the `arrow` crate which use `unsafe` and `transmute` code internally. We are actively working as a community to minimize undefined behavior and remove `unsafe` usage to align more with Rust's core principles of safety (e.g. the arrow2 project).
+
+As `arrow` exists today, it is fairly easy to misuse the APIs, leading to undefined behavior, and it is especially easy to misuse code in modules named above. For an example, as described in [the arrow2 crate](https://github.com/jorgecarleitao/arrow2#why), the following code compiles, does not panic, but results in undefined behavior:
```
Comment (paddyhoran) [inline arrow/README.md:50]: "it is fairly easy to misuse the APIs" - isn't this mostly a statement about `ArrayData`?  Do many people encounter UB using the other API's in Arrow?

What do you think about putting `ArrayData` behind the `unsafe` flag also?

Comment (alamb) [reply arrow/README.md:50]: I agree that if code is using `ArrayData` directly it is likely to be of the "easy to misuse category". What do others think?

Comment (alamb) [reply arrow/README.md:50]: @paddyhoran  I am going to merge this PR and perhaps we can clarify / improve the wording about `ArrayData` in a follow on PR?

**Pull request merged**